chromium_src/base/json/json_reader.cc

```diff
@@ -0,0 +1,17 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "src/base/json/json_reader.cc"
+
+namespace base {
+
+#if BUILDFLAG(BUILD_RUST_JSON_READER)
+JSONReader::Result DecodeJSONInRust(std::string_view json, int options) {
+  SCOPED_UMA_HISTOGRAM_TIMER_MICROS(kSecurityJsonParsingTime);
+  return DecodeJSONInRust(json, options, internal::kAbsoluteMaxDepth);
+}
+#endif  // BUILDFLAG(BUILD_RUST_JSON_READER)
```

Comment:
It looks like this slipped through review, but this should not have been added to our overrides, as this is not much different of just using the official API `JSONReader::Read`/`JSONReader::ReadAndReturnValueWithError`, specially as this particular override is only used by `APIRequestResult`. We shouldn't be introducing unneeded overrides.

This looks innocuous but I had to build brave with a custom tool and I'm now hitting build errors because of how this override introduces a dependency to a buildflag gen file:

```
../../brave/chromium_src/base/json/json_reader.h:8:10: fatal error: 'base/rust_buildflags.h' file not found
    8 | #include "base/rust_buildflags.h"       // IWYU pragma: private
      |          ^~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```

In general **only use overrides when absolutely necessary**.

cc @atuchin-m @bridiver @iefremov @goodov

Reply:
I've just looked at this more carefully. It looks like the parser code was like this:
```cxx
// static
JSONReader::Result JSONReader::ReadAndReturnValueWithError(
    std::string_view json,
    int options) {
#if BUILDFLAG(BUILD_RUST_JSON_READER)
  SCOPED_UMA_HISTOGRAM_TIMER_MICROS(kSecurityJsonParsingTime);
  if (UsingRust()) {
    return DecodeJSONInRust(json, options, internal::kAbsoluteMaxDepth);
  } else {
    internal::JSONParser parser(options);
    auto value = parser.Parse(json);
    if (!value) {
      Error error;
      error.message = parser.GetErrorMessage();
      error.line = parser.error_line();
      error.column = parser.error_column();
      return base::unexpected(std::move(error));
    }

    return std::move(*value);
  }
#else   // BUILDFLAG(BUILD_RUST_JSON_READER)
  internal::JSONParser parser(options);
  auto value = parser.Parse(json);
  if (!value) {
    Error error;
    error.message = parser.GetErrorMessage();
    error.line = parser.error_line();
    error.column = parser.error_column();
    return base::unexpected(std::move(error));
  }

  return std::move(*value);
#endif  // BUILDFLAG(BUILD_RUST_JSON_READER)
```

Which means this function call was guarded behind `BUILD_RUST_JSON_READER`, however now it looks like this on master:
```
JSONReader::Result JSONReader::ReadAndReturnValueWithError(
    std::string_view json,
    int options) {
#if BUILDFLAG(IS_NACL)
  internal::JSONParser parser(options);
  auto value = parser.Parse(json);
  if (!value) {
    Error error;
    error.message = parser.GetErrorMessage();
    error.line = parser.error_line();
    error.column = parser.error_column();
    return base::unexpected(std::move(error));
  }
  return std::move(*value);
#else   // BUILDFLAG(IS_NACL)
  SCOPED_UMA_HISTOGRAM_TIMER_MICROS(kSecurityJsonParsingTime);
  if (UsingRust()) {
    return DecodeJSONInRust(json, options, internal::kAbsoluteMaxDepth);
  } else {
    internal::JSONParser parser(options);
    auto value = parser.Parse(json);
    if (!value) {
      Error error;
      error.message = parser.GetErrorMessage();
      error.line = parser.error_line();
      error.column = parser.error_column();
      return base::unexpected(std::move(error));
    }
    return std::move(*value);
  }
#endif  // BUILDFLAG(IS_NACL)
}
```

For things like this we must consider how long an override like this will stay valid, because this is an active project that is being switched on. I'm not sure if anyone reached out to chromium people to check with them if they would allow some interface to be exposed, so the reversal would have become a bit more obvious once the rust parser became default, as it is now.

Reply:
Looking at this a bit further 😄 and understand how this properly works, and how brave consumes this, I'm now convinced we should have just switched `BUILD_RUST_JSON_READER`, and have dispensed of the overrides in the first place.

Reply:
@cdesouza-chromium Using Rust in JSONReader is controlled by 1) `BUILD_RUST_JSON_READER` flag 2) base::Feature [UseRustJsonParser](https://source.chromium.org/chromium/chromium/src/+/main:base/features.cc;drc=df39aeaf525d75ccb31ad3ca0a52eb1ba5a64451;l=45).
We enabled the build flag (and now it's also enabled in chromium), but we didn't want to touch the feature (because it would affect every JSONReader)
That's why we export this function to use `DecodeJSONInRust` directly.

Currently, the feature also enabled for all the platform expect android.
It's likely Android will also be covered soon: https://issues.chromium.org/issues/40276029
Any particular issue with this code now?

Reply:
Well, I didn't explain myself well, and at first I also didn't understand the problem too well.

There wasn't an issue actually. I was just leaving a few observations. I think we knew that upstream would eventually switch this on, so this could have been written anticipating that flip if the switch, when chromium would just make rust parsing the default. I had to rework the code a bit now that it is the default.

I also think for cases like this, where chromium is already heading a certain direction, we could have discussed with `//base` owners, and got something in chromium upstream as well, rather than relying on an override.

Anyway, thanks mate.